src/queries/text.ts

```diff
@@ -58,37 +59,65 @@ const getChildrenAsText = (
 const getNodeByText = (
   node: ReactTestInstance,
   text: TextMatch,
+  TextComponent: React.ComponentType,
   options: TextMatchOptions = {}
 ) => {
-  try {
-    const { Text } = require('react-native');
-    const isTextComponent = filterNodeByType(node, Text);
-    if (isTextComponent) {
-      const textChildren = getChildrenAsText(node.props.children, Text);
-      if (textChildren) {
-        const textToTest = textChildren.join('');
-        const { exact, normalizer } = options;
-        return matches(text, textToTest, normalizer, exact);
-      }
+  const isTextComponent = filterNodeByType(node, TextComponent);
+  if (isTextComponent) {
+    const textChildren = getChildrenAsText(node.props.children, TextComponent);
+    if (textChildren) {
+      const textToTest = textChildren.join('');
+      const { exact, normalizer } = options;
+      return matches(text, textToTest, normalizer, exact);
     }
-    return false;
-  } catch (error) {
-    throw createLibraryNotSupportedError(error);
   }
+  return false;
 };
 
+function getCompositeParent(
+  element: ReactTestInstance,
+  compositeType: React.ComponentType
+) {
+  if (!isHostElement(element)) return null;
+
+  let current = element.parent;
+  while (!isHostElement(current)) {
+    // We're at the top of the tree
+    if (!current) {
+      return null;
+    }
+
+    if (filterNodeByType(current, compositeType)) {
+      return current;
+    }
+    current = current.parent ?? null;
+  }
+
+  return null;
+}
+
 const queryAllByText = (
   instance: ReactTestInstance
 ): ((
   text: TextMatch,
   options?: TextMatchOptions
 ) => Array<ReactTestInstance>) =>
   function queryAllByTextFn(text, options) {
-    const results = instance.findAll((node) =>
-      getNodeByText(node, text, options)
-    );
+    try {
+      const { Text } = require('react-native');
+      const rootInstance = isHostElement(instance)
+        ? getCompositeParent(instance, Text) ?? instance
```

Comment:
The issue here is that we'll get into this line of the ternary also for non `Text` host components. For instance, when searching `getByRole('button', { name })`, it'll try to find a Text higher in the tree and fail. So we need to default to `?? instance`.

A better (more performant) solution would be to do `const rootInstance = filterNodeByType(instance, 'Text')`, so we'll hit this path only when needed. But we fall back to the hardcoded host string `Text`.

Reply:
What if instead of `isHostElement` you used `isHostTextInstance`?:

```ts
function isHostTextElement(element: ReactTestInstance) {
  // Not a host element
  if (typeof element.type !== "string") return false;
  
  let current = element.parent;
  while(typeof current.type !== 'string') {
    if (current.type === RN.Text) { 
      return true;
    }
    current = current.parent;
  }
  
  return false;
}
```

Reply:
Would it be possible for a host instance that is not `Text` to have a composite parent that is `Text`? Because if that's the case we'd want to avoid going higher up the tree (and take the risk of going to a sibling).

Reply:
Assuming that when traversing up in the component tree there are intermittent host ancestors along the way, only composite ones, then I think we're pretty safe. 

If non-`Text` host instance would have composite ancestor of exported `Text` type (assuming no host ancestors along the way), then it would mean that exported `Text` renders not only host `Text` but also some other host component.

It seems that the above check (`isHostTextElement`) is atm the most reliable method for checking that given host component is a host `Text` that we have.

@thymikee maybe you have some thoughts on this...

Reply:
BTW it seems that this whole problem exists only because of the way we collect component text content. If we would have https://github.com/facebook/react/pull/25329 merged, then we would have a better alternative for getting element text content.

Reply:
I'd appreciate adding a heavy suite of tests with lots of configurations of Text, and non-Text components with all possible configurations and going with @mdjastrzebski solution for now. We'll be able to move to a simple string check once the name of the component is a public API of RN. Which is something open for a discussion in the https://github.com/react-native-community/discussions-and-proposals repo. Anybody would like to create a proposal for at least Text and View host components? :)

Reply:
https://github.com/facebook/react/pull/25329 would be for React 18 only though

Reply:
@thymikee I've submitted PR for such tests https://github.com/callstack/react-native-testing-library/pull/1143

Reply:
Just repushed with a carpet bombing approach on tests. Let me know if you feel I should keep/remove them (didn't see your competing PR)